jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt

```diff
@@ -3,146 +3,122 @@
 
 package software.aws.toolkits.jetbrains.ui
 
-import com.intellij.openapi.application.ApplicationManager
 import com.intellij.openapi.application.ModalityState
 import com.intellij.openapi.application.runInEdt
+import com.intellij.openapi.project.Project
 import com.intellij.openapi.ui.ComboBox
-import com.intellij.openapi.ui.ValidationInfo
+import org.jetbrains.annotations.TestOnly
+import software.aws.toolkits.core.utils.Either
 import software.aws.toolkits.core.utils.getLogger
 import software.aws.toolkits.core.utils.warn
-import software.aws.toolkits.jetbrains.utils.ui.selected
+import software.aws.toolkits.jetbrains.core.AwsResourceCache
+import software.aws.toolkits.jetbrains.core.Resource
+import software.aws.toolkits.jetbrains.utils.ui.find
 import software.aws.toolkits.resources.message
 import javax.swing.DefaultComboBoxModel
 
-class ResourceSelector<T> : ComboBox<T>() {
+typealias Selector<T> = Either<Any?, (T) -> Boolean>
 
-    var loadingStatus: ResourceLoadingStatus = ResourceLoadingStatus.SUCCESSFUL
-        private set
+class ResourceSelector<T>(project: Project, private val resourceType: Resource<out Collection<T>>) : ComboBox<T>() {
 
+    private val resourceCache = AwsResourceCache.getInstance(project)
+    private val actualModel = this.model as DefaultComboBoxModel<T>
+    private var loadingStatus: Status = Status.NOT_LOADED
     private var shouldBeEnabled: Boolean = isEnabled
+    private var selector: Selector<T>? = null
 
-    var loadingException: Exception? = null
-        private set
+    init {
+        if (isEnabled) {
+            reload()
+        }
+    }
 
-    /**
-     * Postpone the ComboBox enability changing when it is still in loading status. Ie. when the ComboBox is in
-     * [ResourceLoadingStatus.LOADING]    - always disable it, and change to the desired status after successfully loaded.
-     * [ResourceLoadingStatus.SUCCESSFUL] - same as the standard [setEnabled] behavior.
-     * [ResourceLoadingStatus.FAILED]     - always disable it, and always set the desired status to false.
-     */
-    @Synchronized override fun setEnabled(enabled: Boolean) {
-        shouldBeEnabled = when (loadingStatus) {
-            ResourceLoadingStatus.SUCCESSFUL -> {
-                super.setEnabled(enabled)
-                enabled
-            }
-            ResourceLoadingStatus.LOADING -> {
-                super.setEnabled(false)
-                enabled
-            }
-            else -> {
-                super.setEnabled(false)
-                false
+    @JvmOverloads
+    @Synchronized
+    fun reload(forceFetch: Boolean = false) {
+        val previouslySelected = actualModel.selectedItem
+        loadingStatus = Status.LOADING
+        runInEdt(ModalityState.any()) {
+            super.setEnabled(false)
+            setEditable(true)
```

Comment:
Why is it editable when loading?

Reply:
that's the only way we can get it to update `selectedText` you can't set a non-`T` value unless it's editable - it's disabled so the effect isn't actually seen by the customer.

Reply:
Ah, k